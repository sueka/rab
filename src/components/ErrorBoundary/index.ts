import * as React from 'react'

interface LocalState {
  hasError: boolean
  error?: unknown
}

class ErrorBoundary extends React.Component<unknown, LocalState> {

  // NOTE: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/826ce0f1ce1d1887d199986283630d6f63075ad5/types/react/index.d.ts#L419 にも関わらず、初期化されていない state は null であるため、初期化を強制するためにプロパティ宣言を行う。
  public state: Readonly<LocalState> = {
    hasError: false,
  }

  public static getDerivedStateFromError = (error: unknown) => ({
    hasError: true,
    error,
  })

  public render() {
    const { children } = this.props
    const { hasError, error } = this.state

    if (hasError) {
      if (error instanceof Error) {
        return `${ error }`
      }

      throw new TypeError(`${ error } is not an error.`)
    }

    return children
  }
}

export default ErrorBoundary
