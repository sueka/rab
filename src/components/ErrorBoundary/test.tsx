import * as React from 'react'
import { render } from '@testing-library/react'

import ErrorBoundary from '.'

const NoThrowing: React.FunctionComponent = () => <>no throwing</>
const ErrorThrowing: React.FunctionComponent = () => { throw new Error('error throwing') }
const NonErrorThrowing: React.FunctionComponent = () => { throw 'non-error throwing' }

describe('ErrorBoundary', () => {
  test('without anything throwing', () => {
    const { container } = render(
      <ErrorBoundary>
        <NoThrowing />
      </ErrorBoundary>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  test('with an error throwing', () => {
    const { container } = render(
      <ErrorBoundary>
        <ErrorThrowing />
      </ErrorBoundary>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  test('with a non-error throwing', () => {
    expect(() => {
      render(
        <ErrorBoundary>
          <NonErrorThrowing />
        </ErrorBoundary>
      )
    }).toThrowError()
  })
})
