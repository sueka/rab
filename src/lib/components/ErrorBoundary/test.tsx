import React from 'react'
import { render } from '@testing-library/react'

import ErrorBoundary from '.'

const NoErrorThrowing: React.FunctionComponent = () => <>no error throwing</>
const ErrorThrowing: React.FunctionComponent = () => { throw new Error('error throwing') }
const NonErrorThrowing: React.FunctionComponent = () => {
  throw 'non-error throwing' // tslint:disable-line:no-string-throw
}

describe('ErrorBoundary', () => {
  test('without anything throwing', () => {
    const renderError = jest.fn()

    const { container } = render(
      <ErrorBoundary renderError={ renderError }>
        <NoErrorThrowing />
      </ErrorBoundary>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  test('with an error throwing', () => {
    const renderError = jest.fn()

    const { container } = render(
      <ErrorBoundary renderError={ renderError }>
        <ErrorThrowing />
      </ErrorBoundary>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  test('with a non-error throwing', () => {
    const renderError = jest.fn()

    expect(() => {
      render(
      <ErrorBoundary renderError={ renderError }>
          <NonErrorThrowing />
        </ErrorBoundary>
      )
    }).toThrowError()
  })
})
