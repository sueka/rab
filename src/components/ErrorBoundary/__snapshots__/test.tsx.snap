// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary with an error throwing 1`] = `Error: error throwing`;

exports[`ErrorBoundary without anything throwing 1`] = `no throwing`;
